### PR TITLE
feat(utils): allow specifying additional rule meta.docs in RuleCreator

### LINKED ISSUE #8695: Bug: RuleCreator leaks internal meta.docs, meta.type types
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Playground Link

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgUQMoBlgDsYFUbAA2AznAL5wBmUEIcA5AAIwCeYApsQMZTBgwBaToWwwA9AFcCJegG4AUPPYAPSLDhcIWYvCgTC7OAF4UGUfiLEAdACV97AMJR2AQxjQrAd2AwAFhCkAEQguYgAKBHk4DWc3djDNHBUYAEpEKOi4ZxgJKCx0zMyASQATdhxgSmB2KDCsCDK0yMLCxJhkq2c1GAi4EE5iFwBzdlKALjgAIhUXcANJgBo4erLyFIUW8gzosgyyBYyyyhd9GAB5fmAtYgmAbQBdA+j+mBcJ5sySkJuClrLuXiXLQTej0J4tZyaED9LBlEogyG0GFwsHbTIQPw1CYwPTscE7fF9AbDTjvNHRGZzdgTSZnXzLCAAQkmaP2aNYHBBhBcLACMFRhW4vnYIDecAeewOZHWiiAA

### Repro Code

```TypeScript
import { ESLintUtils } from '@typescript-eslint/utils';

export const rule = ESLintUtils.RuleCreator.withoutDocs({
  create(context) {
    return {
      Identifier(node) {
        context.report({ messageId: "example", node });
      }
    }
  },
  defaultOptions: [],
  meta: {
    docs: {
      description: '',
      recommended: 'recommended',
      other: true,
    },
    messages: {
      example: "Oh no!"
    },
    type: 'layout',
    schema: []
  },
});
```

### ESLint Config

```javascript
n/a
```

### tsconfig

```jsonc
n/a
```

### Expected Result

Custom rules in community plugins shouldn't be tied into our types for what goes into a rule's `meta.docs`. Plugins might define their own, potentially different documentation metadata & strategies.

Maybe we should make `meta.docs` type... `Record<string, unknown>` for consumers? Or a type parameter on `RuleCreator` that defaults to `Record<string, unknown>`?

### Actual Result

* Known properties have types set to ours: e.g. `meta.docs.recommended` is `RuleRecommended | undefined` in 7.2.0
    * After #8364 (7.3.0), it'll be `RuleRecommendation | RuleRecommendationAcrossConfigs<Options> | undefined`
* Unknown properties cause a type error: e.g. `meta.docs.other`

### Additional Info

See also:

* https://github.com/eslint/eslint/issues/17842: for a more general approach in ESLint core
* https://github.com/typescript-eslint/typescript-eslint/pull/8364#discussion_r1527017534: noting that #8364's expansion of the `meta.docs.recommended` property will be visible to users

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/utils`             | `7.2.0` |

💖

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
BREAKING CHANGE:

Changes the default allowed types on custom rule docs.

## PR Checklist

- [x] Addresses an existing open issue: fixes #8695
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds an optional `PluginDocs` type parameter to `RuleCreator` that allows adding additional properties to `meta.docs`. That allows us to move `extendsBaseRule`, `recommended`, and `requiresTypeChecking` into our own `ESLintPluginDocs` interface so that they're no longer exposed to consumers.

Adds new `RuleModuleWithMetaDocs` and `AnyRuleModuleWithMetaDocs` interfaces so that we can finally get rid of the `?.` after `meta.docs` in some of our internal tooling. And `RuleMetaDataDocs` no longer has an `Options` parameter, which simplifies some other types too.

Adds back in `"layout"` to `meta.type` as it actually is specified in https://eslint.org/docs/latest/extend/custom-rules.

💖

### CODE REVIEW

File: packages/eslint-plugin/rules.d.ts
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/rules.d.ts:1]: cc @JamesHenry, I'm very confused :grim

[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/eslint-plugin/rules.d.ts:1]: @JoshuaKGoldberg the website project does depend on the eslint-plugin (package.json and numerous import statements) and the fix seems reasonable:

image

image

[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/eslint-plugin/rules.d.ts:1]: Shall we just go ahead and fix it here?

[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/eslint-plugin/rules.d.ts:1]: Plus things will shortly become a lot more transparent when it comes to resolving types in the typecheck command (per my discord message earlier)

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/rules.d.ts:1]: Makes sense, might as well fix it yeah. I was more confused as to why the error was only happening now - and not before.

**Review by bradzacher — APPROVED** [review]
nodyes

**Pull request merged**